theseus/optimizer/nonlinear/dogleg.py

```diff
@@ -0,0 +1,121 @@
+        delta_gn = self.linear_solver.solve()
+        delta_gn_norm_2 = TrustRegion._squared_norm(delta_gn)
+        good_gn_idx = delta_gn_norm_2 < trust_region_2
+        # All Gauss-Newton step are within trust-region, can return
+        if good_gn_idx.all():
+            # Return a False mask since no indices are at the boundary
+            return delta_gn, torch.zeros_like(good_gn_idx).bool()
+
+        # ---------------------------------------------------------------------
+        # Some Gauss-Newton steps are outside trust-region,
+        # need to compute dogleg step
+        # ---------------------------------------------------------------------
+
+        linearization = self.linear_solver.linearization
+
+        neg_grad = linearization.Atb.squeeze(2)
+        Adelta_sd = linearization.Av(neg_grad)
+        Adelta_sd_norm_2 = TrustRegion._squared_norm(Adelta_sd)
+        grad_norm_2 = TrustRegion._squared_norm(neg_grad)
+        t = grad_norm_2 / (Adelta_sd_norm_2 + Dogleg.EPS)
+        delta_sd = neg_grad * t
+        delta_sd_norm_2 = TrustRegion._squared_norm(delta_sd)
+        not_near_zero_sd_idx = delta_sd_norm_2 > 1e-6
+        sd_within_region_idx = delta_sd_norm_2 <= trust_region_2
+
+        # First make sure that any steps beyond the trust region, are truncated
+        if not sd_within_region_idx.all():
+            delta_dogleg = torch.where(
+                sd_within_region_idx,
+                delta_sd,
+                delta_sd * self._trust_region / (delta_sd_norm_2 + Dogleg.EPS).sqrt(),
+            )
+        else:
+            delta_dogleg = delta_sd
+
+        # Now mask near zero indices so the next computation doesn't happen for them
+        sd_within_region_idx = sd_within_region_idx & not_near_zero_sd_idx
+
+        if sd_within_region_idx.any():
+            # In this case, some steepest descent steps are within region
+            # so need to extend towards boundary with Gauss-Newton step
+            # Need to solve a quadratic || sd + tau * (gn - sd)|| == tr**2
+            # This can be written as
+            # a * tau^2 + b * tau + c, with a, b, c given below
+            diff = delta_gn - delta_sd
+            a = TrustRegion._squared_norm(diff)
+            b = (2 * delta_sd * diff).sum(dim=1, keepdim=True)
+            c = delta_sd_norm_2 - trust_region_2
+            disc = ((b**2) - 4 * a * c).clamp(0.0)
+            # By taking min(tau, 1), this also covers the case when ||d_gn|| < TR
+            tau = ((-b + disc.sqrt()) / (2 * a + Dogleg.EPS)).minimum(disc.new_ones(1))
+            delta_dogleg = torch.where(
+                sd_within_region_idx,
+                delta_sd + tau * diff,
+                delta_dogleg,
+            )
+
+        # Finally, when the steepest descent direction is too close to zero, just use
+        # the Gauss-Newton direction truncated at the trust region, to avoid
+        # numerical errors.
+        if not not_near_zero_sd_idx.all():
```

Comment:
What will happen if `||delta_gn||<self._trust_region`? Do you have a reference for this step?

Reply:
Mmmm, I actually think this is a bug. There was a line I removed that filtered for that case so that we would choose the GN step. Maybe we don't need this at all if we do the right thing when computing `tau`. I'll take a look.